### DISCUSSION TITLE
Error when execute tests/benchmarks/polybench

### DISCUSSION THREAD

**Discussion by jquanC** [bytecodealliance/wasm-micro-runtime#2186] [Q&A] [answered]
I follow the touriol in [tests/benchmarks/polybench](https://github.com/bytecodealliance/wasm-micro-runtime/tree/main/tests/benchmarks/polybench) to execuete the polybench. Before this, I think I have successfully finished the preliminary (build iwasm and wamrc AOT compiler) because I have executed some basic samples.

After that I run `./build.sh` in the command line, and received one error as follows:
>...
1 error generated.
Compile seidel-2d.wasm into seidel-2d.aot
Read file to buffer failed: open file /opt/wasm-micro-runtime/tests/benchmarks/polybench/out/seidel-2d.wasm failed.
Build adi_native
Build adi.wasm
In file included from utilities/polybench.c:19:
/opt/wasi-sdk/bin/../share/wasi-sysroot/include/sys/resource.h:2:2: error: WASI lacks process-associated clocks; to enable emulation of the `getrusage` function using the wall clock, which isn't sensitive to whether the program is running or suspended, compile with -D_WASI_EMULATED_PROCESS_CLOCKS and link with -lwasi-emulated-process-clocks
#error WASI lacks process-associated clocks; to enable emulation of the `getrusage` function using \
 ^
1 error generated.
Compile adi.wasm into adi.aot
Read file to buffer failed: open file /opt/wasm-micro-runtime/tests/benchmarks/polybench/out/adi.wasm failed.
Done

I get a little more information about it from [libc-top-half/musl/include/sys/resource.h](https://github.com/WebAssembly/wasi-libc/blob/main/libc-top-half/musl/include/sys/resource.h) and [commit info](https://github.com/WebAssembly/wasi-libc/commit/b9b64a695a35273f92574acc6270f0e059be3dca)

since the committer says:
>Emulate clock, times, and getrusage using the monotonic clock.
Upcoming WASI snapshots omit the `PROCESS_CPUTIME` clock, since WASI has
no inherent concept of processes, and since implementations which don't
use a process for each instance don't have a way to implement it
efficiently.
However, `clock`, `times`, and `getrusage` are useful functions, so
provide optional emulated version of them, using the `MONOTONIC` clock.
This means these implementations will measure not just the program's
own CPU time, but also time spent suspended while other programs are
running.
Due to this difference in behavior, put these implementations behind
a flag. Users must pass `-D_WASI_EMULATED_PROCESS_CLOCK` and link with
`-lwasi-emulated-process-clocks` to enable them.

I am still not sure how to solve this error. I am not familiar with the building and development of C/C++ programs (currently making up).
And my understanding is:
1. If I ignore this error, I can continue the sample by running the offered bash file to test the benchmark, while the time spent in the result file may not be accurate.
2. If I want to use function like `clock`, `times`, and `getrusage` to get an accurate report, I have to solve the error.

I want to know if my understanding is right or not.  And how to eliminate this error? （Should I need to modify the` ./build.sh `?）
Does someone knows and help me？

Env: ubuntu 18.04
WASI version: wasi-sdk-19

**Comment by wenyongh** [comment, marked as answer]
Hi, please add -D_WASI_EMULATED_PROCESS_CLOCKS when using wais-sdk-19.0 to build the polybench benchmark, I submitted PR #2187 to fix it.

**Reply by jquanC** [reply]
It works, thx!
